sympy/solvers/ode/ode.py

```diff
@@ -1845,39 +1829,27 @@ def odesimp(ode, eq, func, hint):
     if eq.lhs == func and not eq.rhs.has(func):
         # The solution is already solved
         eq = [eq]
-
-        # special simplification of the rhs
         if hint.startswith("nth_linear_constant_coeff"):
-            # Collect terms to make the solution look nice.
-            # This is also necessary for constantsimp to remove unnecessary
-            # terms from the particular solution from variation of parameters
-            #
-            # Collect is not behaving reliably here.  The results for
-            # some linear constant-coefficient equations with repeated
-            # roots do not properly simplify all constants sometimes.
-            # 'collectterms' gives different orders sometimes, and results
-            # differ in collect based on that order.  The
-            # sort-reverse trick fixes things, but may fail in the
-            # future. In addition, collect is splitting exponentials with
-            # rational powers for no reason.  We have to do a match
-            # to fix this using Wilds.
-            #
-            # XXX: This global collectterms hack should be removed.
-            global collectterms
-            collectterms.sort(key=default_sort_key)
-            collectterms.reverse()
-            assert len(eq) == 1 and eq[0].lhs == f(x)
             sol = eq[0].rhs
             sol = expand_mul(sol)
-            for i, reroot, imroot in collectterms:
-                sol = collect(sol, x**i*exp(reroot*x)*sin(abs(imroot)*x))
-                sol = collect(sol, x**i*exp(reroot*x)*cos(imroot*x))
-            for i, reroot, imroot in collectterms:
-                sol = collect(sol, x**i*exp(reroot*x))
-            del collectterms
-
-            # Collect is splitting exponentials with rational powers for
-            # no reason.  We call powsimp to fix.
+            arr=eq[0].rhs.atoms(exp)
+            arr2=eq[0].rhs.atoms(sin)
+            arr3=eq[0].rhs.atoms(cos)
```

Comment:
Can we just do `rhs.atoms(exp, sin, cos)` and then collect on that?

Reply:
The reason I did separately was because we also want to do `collect(expr, exp(x)*sin(x))` basically on all combinations of `exp*sin` and `exp*cos`

Reply:
Does it make a different to the end result?

Reply:
The ans is correct without it also, but it was combining some examples in a nicer way. If you want me to remove, I will do that

Reply:
When you say "in a nicer way" it's hard for me to know what you mean without any examples.

Reply:
My bad. It was related to ordering only. 
I tried with `rhs.atoms(exp, sin, cos)` and ran collect over it. It change solution like:
```
The ODE is:
f(x) - Derivative(f(x), x) + Derivative(f(x), (x, 2))

The expected solution was
Eq(f(x), (C1*sin(sqrt(3)*x/2) + C2*cos(sqrt(3)*x/2))*exp(x/2))

What dsolve returned is:
Eq(f(x), (C1*sin(sqrt(3)*x/2) + C2*cos(sqrt(3)*x/2))*sqrt(exp(x)))

```

Reply:
Why does that come out different? There must be something else turning `exp(x/2)` into `sqrt(exp(x))` because `collect` shouldn't do that.

Reply:
The reason was as `.atoms()` returns a set so it was returning `{cos(sqrt(3)*x/2), sin(sqrt(3)*x/2), exp(x/2)}` for the above mentioned example whereas we wanted to collect first on exp terms. That's why I have used a list now.

Reply:
What happens if you use the `simpsol` function from `systems.py` here? Does it give different results?

Reply:
As far as I understood, simpsol takes list of symbols to be collected on but we want this collect to be on given functions i.e exp,sin and cos.
Although I tried by passing constants like `simpsol(eq, [C1], constants)`, here is the example of simple equations: 
```python
>>> eq=f(x).diff(x, 3) - f(x).diff(x) - sinh(x) - exp(x)
>>> dsolve(eq)
[Eq(f(x), C1*exp(-x) + C2 + C3*exp(x) + x*(exp(x) + sinh(x))/2)]   #solution returned by simpsol
Eq(f(x), C1 + C3*exp(-x) + x*sinh(x)/2 + (C2 + x/2)*exp(x))   #expected solution

>>> eq=f(x).diff(x, 2) - f(x) - exp(x - 1)
>>> dsolve(eq)
[Eq(f(x), C1*exp(-x) + C2*exp(x) + x*exp(x - 1)/2)]  #solution returned by simpsol
Eq(f(x), C2*exp(-x) + (C1 + x*exp(-1)/2)*exp(x))  #expected solution
```

Reply:
The idea with `simpsol` is that you tell it to collect like `simpsol(eq, [C1, C2, ...], [t])` and then it understands that `t` could mean `exp(t)`, `sin(t)*exp(t)` etc. It finds everything that involves `t`. You can also tell it to collect like `simpsol(eq, [t], [C1, C2, ...])` to reverse the order between collecting on `t` vs collecting on the constants.

Reply:
I tried `simpsol(eq, [x], constants)`  as well as `simpsol(eq, [x], constants)` . But for some of the examples it doesn't collect properly. Two of them are:
```
The ODE is:
-exp(x) - sinh(x) - Derivative(f(x), x) + Derivative(f(x), (x, 3))

The expected solution was
[Eq(f(x), C1 + C3*exp(-x) + x*sinh(x)/2 + (C2 + x/2)*exp(x))]

What dsolve returned is:
Eq(f(x), C1 + C2*exp(-x) + C3*exp(x) + x*exp(x)/2 + x*sinh(x)/2)

```

```
The ODE is:
-x - x*exp(-x) - f(x) + Derivative(f(x), x) + 5*Derivative(f(x), (x, 2)) + 3*Derivative(f(x), (x, 3))

The expected solution was
[Eq(f(x), C3*exp(x/3) - x + (C1 + C2*x - x**3/24 - 3*x**2/32)*exp(-x) - 1)]

What dsolve returned is:
Eq(f(x), C1*x*exp(-x) + C2*exp(-x) + C3*exp(x/3) - x**3*exp(-x)/24 - 3*x**2*exp(-x)/32 - x - 1)

```

Reply:
Maybe `simpsol` can be improved. I'm still not sure which of these forms I prefer. I think that this is something good to discuss in the meeting. Can you collect examples of how solutions look when collected in different ways so that we can compare them in the meeting?

Reply:
okay, sure I will write them.

Reply:
Here is the list of examples. [Link](https://docs.google.com/document/d/1i4Zr22VXeDRA0eialHRlLmp-2zbp_ALUdGr21ceOnd4/edit?usp=sharing)

Reply:
Which way round are you passing the arguments to `simpsol`? Does that make a difference? How is `simpsol` called in `systems.py`?

Reply:
These examples in the doc were generated using `simpsol(eq, constants, [x])` . In most of the examples it wasn't making any difference when simpsol was called like this `simpsol(eq, [x],  constants)`.  In this [comment](https://github.com/sympy/sympy/pull/21654#discussion_r659735157) only these two examples doesn't change when the order of arguments was swapped.

Reply:
In `systems.py` simpsol is called like `simpsol(sol, [t], constants, doit=doit)` where t is Independent variable in the system of ODEs